attributes-next.md

```diff
@@ -0,0 +1,252 @@
+---
+RFC: unassigned
+Title: Attributes v.Next
+Author: Lamont Granquist <lamont@chef.io>
+Status: Draft
+Type: <Standards Track>
+---
+
+# Attributes v.Next
+
+Overhauling Attributes to be simpler.
+
+## Motivation
+
+    As a Chef User,
+    I want Attributes to be Less Confusing.
+
+    As a Chef User,
+    I dislike NoMethodError for NilClass Errors.
+
+    As a Chef Developer,
+    The monolithic Node attributes API tries to do too much with a shared namespace.
+
+    As a Maintainer of Existing Cookbooks,
+    I need a slow and backwards compatibile transition to any new API.
+
+    As a Ohai Plugin Developer,
+    I want to be able to add ohai attributes without clobbering someone's cookbook attributes.
+
+    As a Chef Support Person,
+    We see normal attribute persistence cause consistent user confusion.
+
+    As a Chef Support Person,
+    We see array merging (sometimes) to cause consistent user confusion, with most users being surprised by merging behavior.
+
+    As a Chef Support Person,
+    We see consistent user confusion over what the point of force_default is.
+
+    As a Chef Developer,
+    I've never seen anyone use or understand the role/environment precedence order swap between default and override.
+
+## Overall Design
+
+Migration Strategy:
+
+- Existing APIs remain unchanged
+- Deprecation warnings will be emitted for edge cases where the APIs differ
+- Ater fixing the deprecation warnings, autofixing from the old APIs to the new APIs should be possible
+
+Three Separated Namespaces:
+
+- Ohai attributes
+- Default and override attributes
+- Normal/storage attributes
+
+Deprecated Behavior:
+
+- Array deep merging is removed (must be manually replaced with hash-key merging)
+- Deep merging of normal values with default and override is removed (users who were incorrectly using normal
+  as a third precedence level will need to manually fix)
+- Deep merging of automatic values with default/normal/override is removed (this is likely just namespace
+  collision mistakes and stright-up bugs?)
+- The force_default/force_override levels are now the standard behavior of setting attributes in recipe code, the
```

Comment:
You say that, but do you mean `force_default` or `force_default!`?

The piece that was missing when we added these was:

```
node.default['foo']['bar'] = 'stuff'
# I can clear it!
node.override['foo'] = {}
node['foo']
=> {'bar' => 'stuff'}
# Fuck, what about this:
node.force_override['foo'] = {}
node['foo']
=> {'bar' => 'stuff'}
# Damnit! Let's try this!
node.force_override!['foo'] = {}
node['foo']
=> {}
# finally!
```

So if you drop force_default and force_override, you need to either **keep** force_default! and force_override! which sucks, or make `force_default!` and `force_override!` the default.

As a reminder the difference is:
* force_x - Nuke all instances in X precedence (roles_x, cookbooks_x, environments_x)
* force_x! - Nuke all instances in X precedence *and* all lesser precedences.

Please do not break our model again.